### PR TITLE
[diag] add radio sleep / receive mode for RF certification

### PR DESCRIPTION
**Pull request by yuzhyang**
Bug: 127894959

### CODE REVIEW

File: src/diag/diag_process.cpp [hunk 1]
```diff
@@ -307,6 +308,53 @@ void Diag::ProcessStats(int aArgCount, char *aArgVector[], char *aOutput, size_t
     AppendErrorResult(error, aOutput, aOutputMaxLen);
 }
 
+void Diag::ProcessReceive(int aArgCount, char *aArgVector[], char *aOutput, size_t aOutputMaxLen)
```
Comment (bukepo) [inline src/diag/diag_process.cpp]: Make receive/sleep as sub commands of `radio`?
USAGE:
```sh
> diag radio sleep
> diag radio receive
```

Comment (yuzhyang) [reply src/diag/diag_process.cpp]: In README.md there is "diag sleep" but not implemented. From the usage: "Enter radio sleep mode.", actually this command is just what we need here. Do we need to follow the command style in README, then update the README accordingly?

Comment (chshu) [reply src/diag/diag_process.cpp]: @yuzhyang The “diag sleep” cmd was implemented, but removed by #2378 to make it as a platform specific feature, but I forgot to delete it from README. 
Please help to delete the "sleep" cmd, and add "radio sleep/receive" cmd.

Comment (yuzhyang) [reply src/diag/diag_process.cpp]: The commands "send, power, channel, repeat" all seems to be the radio behaviors. Is this OK to keep it short and the function could be single as well?

Comment (chshu) [reply src/diag/diag_process.cpp]: The "diag sleep" may leads to some misunderstanding: radio sleep or MCU sleep (or even chip sleep while radio on RX), that's why we remove the "diag sleep" feature to be platform specific. "diag radio sleep" would be better to describe the behavior.

File: src/diag/diag_process.cpp [hunk 2]
```diff
@@ -45,8 +45,9 @@ namespace ot {
 namespace Diagnostics {
 
 const struct Diag::Command Diag::sCommands[] = {
```
Comment (bukepo) [inline src/diag/diag_process.cpp]: I think we don't need the last `{NULL, NULL}` if we use `OT_ARRAY_LENTH()` to compute the number of items. Would you like to submit another PR to enhance this?

File: src/diag/diag_process.cpp [hunk 3]
```diff
@@ -307,6 +308,51 @@ void Diag::ProcessStats(int aArgCount, char *aArgVector[], char *aOutput, size_t
...
+        otPlatDiagChannelSet(sChannel);
+
+        snprintf(aOutput, aOutputMaxLen, "set radio from sleep to receive on channel %d\r\nstatus 0x%02x\r\n", sChannel,
+                 error);
+    }
+    else
+        error = OT_ERROR_INVALID_ARGS;
+
+exit:
+    OT_UNUSED_VARIABLE(aArgCount);
```
Comment (bukepo) [inline src/diag/diag_process.cpp]: this is used.

File: src/diag/diag_process.cpp [hunk 4]
```diff
@@ -307,6 +308,51 @@ void Diag::ProcessStats(int aArgCount, char *aArgVector[], char *aOutput, size_t
...
+    OT_UNUSED_VARIABLE(aArgVector);
```
Comment (bukepo) [inline src/diag/diag_process.cpp]: this is used.

File: src/diag/diag_process.cpp [hunk 5]
```diff
@@ -307,6 +308,51 @@ void Diag::ProcessStats(int aArgCount, char *aArgVector[], char *aOutput, size_t
...
+            sChannel = static_cast<uint8_t>(value);
+        }
+        error = otPlatRadioReceive(sInstance, sChannel);
+        otPlatDiagChannelSet(sChannel);
+
+        snprintf(aOutput, aOutputMaxLen, "set radio from sleep to receive on channel %d\r\nstatus 0x%02x\r\n", sChannel,
+                 error);
+    }
+    else
+        error = OT_ERROR_INVALID_ARGS;
```
Comment (bukepo) [inline src/diag/diag_process.cpp]: initialize `error` to `OT_ERROR_INVALID_ARGS` and save this line.

File: src/diag/diag_process.cpp (hunk 2 above)
Comment (yuzhyang) [reply src/diag/diag_process.cpp]: Sure, will update in another PR.

File: src/diag/diag_process.cpp [hunk 6]
```diff
@@ -309,6 +310,47 @@ void Diag::ProcessStats(int aArgCount, char *aArgVector[], char *aOutput, size_t
...
+            snprintf(aOutput, aOutputMaxLen, "set radio from receive to sleep \r\nstatus 0x%02x\r\n", error);
+        }
+    }
+    else if (strcmp(aArgVector[0], "receive") == 0)
+    {
+        if (aArgCount > 1)
+        {
+            long value;
+
+            SuccessOrExit(error = ParseLong(aArgVector[1], value));
```
Comment (bukepo) [inline src/diag/diag_process.cpp]: We already have `diag channel`, I suggest keep using that. Would you like submit another PR to add the missing documents for `diag channel`!

File: src/diag/README.md [hunk 7]
```diff
@@ -13,7 +13,8 @@ The diagnostics module supports common diagnostics features that are listed belo
 * [diag power](#diag-power)
 * [diag send](#diag-send-packets-length)
 * [diag repeat](#diag-repeat-delay-length)
-* [diag sleep](#diag-sleep)
+* [diag radio sleep](#diag-radio-sleep)
+* [diag radio receive](#diag-radio-receive)
```
Comment (bukepo) [inline src/diag/README.md]: Just list radio. `sleep` and `receive` should be arguments.

File: src/diag/diag_process.cpp (hunk 6 above)
Comment (yuzhyang) [reply src/diag/diag_process.cpp]: It could be.. But "diag channel" return void; It may need to change diag channel function. Is that OK?

File: src/diag/diag_process.cpp [hunk 8]
```diff
@@ -309,6 +310,41 @@ void Diag::ProcessStats(int aArgCount, char *aArgVector[], char *aOutput, size_t
...
+void Diag::ProcessRadio(int aArgCount, char *aArgVector[], char *aOutput, size_t aOutputMaxLen)
+{
+    otError error = OT_ERROR_INVALID_ARGS;
+
+    VerifyOrExit(otPlatDiagModeGet(), error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(aArgCount > 0, error = OT_ERROR_INVALID_ARGS);
+
+    if (strcmp(aArgVector[0], "sleep") == 0)
+    {
+        /*******otPlatRadioGetState is not supported on travis
```
Comment (bukepo) [inline src/diag/diag_process.cpp]: Remove these?

File: src/diag/diag_process.cpp [hunk 9]
```diff
@@ -309,6 +310,41 @@ void Diag::ProcessStats(int aArgCount, char *aArgVector[], char *aOutput, size_t
...
+                snprintf(aOutput, aOutputMaxLen,
+                         "radio is in %d mode now\r\nset radio from receive to sleep \r\nstatus 0x%02x\r\n",
+                         static_cast<int>(state), error);
+        */
+        error = otPlatRadioSleep(sInstance);
+        snprintf(aOutput, aOutputMaxLen, "set radio from receive to sleep \r\nstatus 0x%02x\r\n", error);
+    }
+    else if (strcmp(aArgVector[0], "receive") == 0)
+    {
+        error = otPlatRadioReceive(sInstance, sChannel);
```
Comment (bukepo) [inline src/diag/diag_process.cpp]: SuccessOrExit?

File: src/diag/diag_process.cpp [hunk 10]
```diff
@@ -309,6 +310,41 @@ void Diag::ProcessStats(int aArgCount, char *aArgVector[], char *aOutput, size_t
...
+                uint8_t state;
+
+                state = otPlatRadioGetState(sInstance);
+
+                error = otPlatRadioSleep(sInstance);
+                snprintf(aOutput, aOutputMaxLen,
+                         "radio is in %d mode now\r\nset radio from receive to sleep \r\nstatus 0x%02x\r\n",
+                         static_cast<int>(state), error);
+        */
+        error = otPlatRadioSleep(sInstance);
```
Comment (bukepo) [inline src/diag/diag_process.cpp]: SuccessOrExit?
